drivers/adc/ad7616/iio_ad7616.c

```diff
@@ -0,0 +1,327 @@
+/***************************************************************************//**
+* @file   iio_ad7616.c
+* @brief  Implementation of AD7616 IIO interface
+********************************************************************************
+* Copyright (c) 2023 Analog Devices, Inc.
+* Copyright (c) 2023 BayLibre, SAS.
+* All rights reserved.
+*
+* This software is proprietary to Analog Devices, Inc. and its licensors.
+* By using this software you agree to the terms of the associated
+* Analog Devices Software License Agreement.
+*******************************************************************************/
+#ifdef IIO_SUPPORT
+
+#include <stdio.h>
+
+#include "ad7616.h"
+#include "iio_ad7616.h"
+#include "iio.h"
+#include "no_os_util.h"
+#include "no_os_alloc.h"
+#include "no_os_error.h"
+
+#define AD7616_CH(_idx) {\
+	.name = "voltage" #_idx, \
+	.ch_type = IIO_VOLTAGE,\
+	.ch_out = IIO_DIRECTION_INPUT,\
+	.indexed = true,\
+	.channel = _idx,\
+	.scan_index = _idx,\
+	.scan_type = &ad7616_iio_scan_type,\
+	.attributes = ad7616_iio_ch_attributes\
+}
+
+#define AD7616_CHN_ATTR(_name, _priv) {\
+	.name = _name,\
+	.priv = _priv,\
+	.show = iio_ad7616_attr_get,\
+}
+
+#define CHN_STORAGE_BITS 16
+
+struct ad7616_iio {
+	struct ad7616_dev *drv_dev;
+	struct iio_device *iio_dev;
+};
+
+enum ad7616_iio_attributes_id {
+	CRC_ATTR_ID,
+};
+
+static int iio_ad7616_attr_get(void *device, char *buf, uint32_t len,
+			       const struct iio_ch_info *channel,
+			       intptr_t priv)
+{
+	struct ad7616_iio_dev *iio_dev = device;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+
+	switch (priv) {
+	case CRC_ATTR_ID:
+		return sprintf(buf, "%u", dev->crc);
+	default:
+		return -EINVAL;
+	}
+}
+
+/**
+ * @brief       Read the debug register value
+ * @param       dev[in, out]- Pointer to IIO device instance
+ * @param       reg[in]- Register address to read from
+ * @param       readval[out]- Pointer to variable to read data into
+ * @return      0 in case of success, negative value otherwise
+ */
+static int32_t iio_ad7616_debug_reg_read(void *device, uint32_t reg,
+		uint32_t *val)
+{
+	struct ad7616_iio_dev *iio_dev = device;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+
+	return ad7616_read(dev,  (uint8_t) reg, (uint16_t *)val);
+}
+
+/**
+ * @brief       Write the debug register value
+ * @param       dev[in, out]- Pointer to IIO device instance
+ * @param       reg[in]- Register address to read from
+ * @param       writeval[out]- Pointer to variable to write data into
+ * @return      0 in case of success, negative value otherwise
+ */
+static int32_t iio_ad7616_debug_reg_write(void *device, uint32_t reg,
+		uint32_t val)
+{
+	struct ad7616_iio_dev *iio_dev = device;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+
+	return ad7616_write(dev, (uint8_t)reg, (uint16_t)val);
+}
+
+void iio_ad7616_event_handler(struct iio_desc *desc)
+{
+	iio_step(desc);
```

Comment:
I don't think iio_step should be called from an iio driver

Reply:
It's not even used anymore, I will remove it